### PR TITLE
smc-fuzzer: init at 0-unstable-2020-12-23

### PR DESCRIPTION
**Pull request by hraban**
## Description of changes
smc-fuzzer "https://github.com/theopolis/smc-fuzzer/" is originally written to fuzz the SMC chip on apple laptops, but it's actually quite useful if you just want to have bona fide interaction with the SMC.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (hraban) [reply pkgs/by-name/sm/smc-fuzzer/package.nix:18]: This software has no releases or tags. Is there a better way to do this? I'm all ears, just thought this was my only option.

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (hraban) [reply pkgs/by-name/sm/smc-fuzzer/package.nix:8]: I'm not against it, just to clarify why I wrote this: this comment is intended for fellow nixpkgs maintainers who might come across this package in the future and ask "why do we have this fuzzer in nixpkgs?" Well, "A: because it's actually useful as a stand-alone tool." But that's not how the tool describes itself, that's just my reason for adding it.

If that kind of editorializing is allowed in longDescription I'm happy to move it there.

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (hraban) [reply pkgs/by-name/sm/smc-fuzzer/package.nix]: Thanks 🤦

File: pkgs/top-level/all-packages.nix
Comment (SigmaSquadron) [inline pkgs/top-level/all-packages.nix:1]: Is this change necessary? Seems a bit off-topic for a package addition.

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (SigmaSquadron) [inline pkgs/by-name/sm/smc-fuzzer/package.nix:1]: Please rename this commit to `smc-fuzzer: init at 0-unstable-2020-12-23` to comply with the [Commit Conventions](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions).

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/sm/smc-fuzzer/package.nix:18]: If this is the case, the `version` attribute must be written as `0-unstable-YYYYMMDD`, where `YYYYMMDD` corresponds to the date of this rev.

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (SigmaSquadron) [inline pkgs/by-name/sm/smc-fuzzer/package.nix:13]:
```suggestion
  version = "0-unstable-2020-12-23";
```
From the [Package Naming Conventions](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming):

> * If a project has no suitable preceding releases - e.g., no versions at all, or an incompatible versioning / tagging schema - then the latest upstream version in the above schema should be 0.
>
>   Example: Given a project that has no tags / released versions at all, or applies versionless tags like latest or YYYY-MM-DD-Build, a commit authored on March 15, 2022 would have version = "0-unstable-2022-03-15".

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (SigmaSquadron) [inline pkgs/by-name/sm/smc-fuzzer/package.nix:34]:
```suggestion
    homepage = "https://github.com/theopolis/smc-fuzzer";
    license = lib.licenses.gpl2Plus;
```
https://github.com/theopolis/smc-fuzzer/blob/ed60c4efeddea8a9778a976c49c650eee3840eac/smc.h#L1-L19

**Review by SigmaSquadron — CHANGES_REQUESTED** [review]
Looks good, though I do have a few nit-picks:

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (SigmaSquadron) [reply pkgs/by-name/sm/smc-fuzzer/package.nix:8]: I don't think there's any hard guidelines w.r.t what's allowed or not in `longDescription`, but something akin to this comment would definitely be relevant.

```nix
meta.longDescription = ''
  `smc-fuzzer` isn’t just useful for fuzzing; it can also be used to explicitly
  communicate with the SMC. This makes it useful in system management,
  e.g. querying / controlling the charge status of an Apple laptop.
'';
```

Comment (AndersonTorres) [reply pkgs/by-name/sm/smc-fuzzer/package.nix:8]: I know of no restrictions on longDescription, except common sense.

File: pkgs/top-level/all-packages.nix
Comment (hraban) [reply pkgs/top-level/all-packages.nix:1]: totally unnecessary I just came across it and thought I'd clean up while I saw it. I created a separate commit for it but I can also create a new PR if that's better.

I used to be stricter about creating separate PRs for every little change but I kinda got told off by other nixpkgs maintainers for it so I'm a bit more lenient now.

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (hraban) [reply pkgs/by-name/sm/smc-fuzzer/package.nix:13]: Thanks I was looking for this but I couldn't find it anywhere. I ended up https://nixos.org/manual/nixpkgs/stable/#sec-using-stdenv here but gave up in the end.

I'll change it 👍

File: pkgs/by-name/sm/smc-fuzzer/package.nix
Comment (hraban) [reply pkgs/by-name/sm/smc-fuzzer/package.nix:34]: Nice! I was looking for that. Thanks, I've also added a PR for a top-level LICENSE file upstream.

**Comment by hraban** [comment]
Thanks for your reviews @AndersonTorres and @SigmaSquadron 👍

File: pkgs/top-level/all-packages.nix
Comment (SigmaSquadron) [reply pkgs/top-level/all-packages.nix:1]: I'm not a committer, so I don't know how they feel about it, but it doesn't make sense to me to bundle completely unrelated changes in one PR. Separating changes makes it easier to review and merge them too, as a minor alphabetical reordering like this one wouldn't need to wait for the main package commit to pass reviews.

**Review by AndersonTorres — APPROVED** [review]
I can't test Apple things, but LGTM

**Review by SigmaSquadron — APPROVED** [review]

File: pkgs/top-level/all-packages.nix
Comment (hraban) [reply pkgs/top-level/all-packages.nix:1]: I'm gonna stick with the feedback I got from the nixpkgs maintainer at the time which was the opposite of this, but if a maintainer comes in and dictates otherwise I'll change my mind.

Tbh I kind of agree with them: it's easier, I'm an unpaid volunteer so fewer PRs = less work, it's also less work to review, there is zero rush for this to get merged so the actual value of "not having to wait" is zero. 🤷  If I'm being honest with myself: at this point if I had to create a separate PR for every little change I'd just stop contributing little changes.

Comment (SigmaSquadron) [reply pkgs/top-level/all-packages.nix:1]: That's fair, and I'd approve either solution anyway. We should really have a CI check for a sorted `all-packages.nix` (or better yet: no `all-packages.nix` at all!) so contributors don't need to worry about it.

Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:1]: I will not complain about  such a minor, cosmetic thing.

However I should  say it is mostly unneeded because:

- The all-packages file will never be in alphabetic order, since the truckload of merges and rebases happening in parallel have no way of enforcing order in such a huge file

- [RFC 140](https://github.com/NixOS/rfcs/pull/140) will make this file obsolete - or at least legacy.
  - Especially the enforcement for future packages: https://github.com/NixOS/nixpkgs/pull/275539
  - Especially the mass migration: https://github.com/NixOS/nixpkgs/pull/211832 will make this file obsolete

Comment (hraban) [reply pkgs/top-level/all-packages.nix:1]: Fair point--removed. Thank you.

Comment (hraban) [reply pkgs/top-level/all-packages.nix:1]: > That's fair, and I'd approve either solution anyway. We should really have a CI check for a sorted `all-packages.nix` (or better yet: no `all-packages.nix` at all!) so contributors don't need to worry about it.

This is also a good point you're right

**Pull request merged**
